include/tscpp/util/Bravo.h

```diff
@@ -0,0 +1,375 @@
+  shared_mutex_impl &operator=(shared_mutex_impl &&) = delete;
+
+  ////
+  // Exclusive locking
+  //
+  void
+  lock()
+  {
+    _mutex.underlying.lock();
+    _revoke();
+  }
+
+  bool
+  try_lock()
+  {
+    bool r = _mutex.underlying.try_lock();
+    if (!r) {
+      return false;
+    }
+
+    _revoke();
+
+    return true;
+  }
+
+  void
+  unlock()
+  {
+    _mutex.underlying.unlock();
+  }
+
+  ////
+  // Shared locking
+  //
+  void
+  lock_shared(Token &token)
+  {
+    // Fast path
+    if (_mutex.read_bias.load(std::memory_order_acquire)) {
+      size_t index = DenseThreadId::self();
+      for (size_t i = 0; i < SLOT_SIZE; ++i) {
+        index       = (index + i) % SLOT_SIZE;
+        Slot &slot  = _mutex.readers[index];
+        bool expect = false;
+        if (slot.mu.compare_exchange_strong(expect, true, std::memory_order_relaxed)) {
+          // recheck
+          if (_mutex.read_bias.load(std::memory_order_acquire)) {
+            token = index + 1;
+            return;
+          } else {
+            slot.mu.store(false, std::memory_order_relaxed);
+          }
+        }
+      }
+    }
+
+    // Slow path
+    _mutex.underlying.lock_shared();
+    if (_mutex.read_bias.load(std::memory_order_acquire) == false && _now() >= _mutex.inhibit_until) {
+      _mutex.read_bias.store(true, std::memory_order_release);
+    }
+  }
+
+  bool
+  try_lock_shared(Token &token)
+  {
+    // Fast path
+    if (_mutex.read_bias.load(std::memory_order_acquire)) {
+      size_t index = DenseThreadId::self();
+      for (size_t i = 0; i < SLOT_SIZE; ++i) {
+        index       = (index + i) % SLOT_SIZE;
+        Slot &slot  = _mutex.readers[index];
+        bool expect = false;
+        if (slot.mu.compare_exchange_weak(expect, true, std::memory_order_release, std::memory_order_relaxed)) {
+          // recheck
+          if (_mutex.read_bias.load(std::memory_order_acquire)) {
+            token = index + 1;
+            return true;
+          } else {
+            slot.mu.store(false, std::memory_order_relaxed);
+          }
+        }
+      }
+    }
+
+    // Slow path
+    bool r = _mutex.underlying.try_lock_shared();
+    if (r) {
+      // Set RBias if the BRAVO policy allows that
+      if (_mutex.read_bias.load(std::memory_order_acquire) == false && _now() >= _mutex.inhibit_until) {
+        _mutex.read_bias.store(true, std::memory_order_release);
+      }
+
+      return true;
+    }
+
+    return false;
+  }
+
+  void
+  unlock_shared(const Token token)
+  {
+    if (token == 0) {
+      _mutex.underlying.unlock_shared();
+      return;
+    }
+
+    Slot &slot = _mutex.readers[token - 1];
+    slot.mu.store(false, std::memory_order_relaxed);
+  }
+
+private:
+  struct alignas(hardware_constructive_interference_size) Slot {
+    std::atomic<bool> mu = false;
+  };
+
+  struct Mutex {
+    std::atomic<bool> read_bias         = false;
+    std::array<Slot, SLOT_SIZE> readers = {};
```

Comment:
It seems unlike the original BRAVO where a global reader table is used, your implementation has a reader table per lock instance. This will increase the per-lock memory footprint, with potential performance benefit. May benchmark the memory usage as well to see whether the trade-off is worthwhile.

Reply:
Right, I followed the approach of xsync's RBMutex. 

https://github.com/puzpuzpuz/xsync/blob/main/rbmutex.go